### PR TITLE
feat: added MacOS menu

### LINKED ISSUE #41: Command Q to quit on Mac OS X
Original: https://github.com/tomaka/glutin/issues/468

Not sure if still relevant.

### PR DESCRIPTION
**Pull request by casperstorm**
This is an WIP PR which shows a way we could add a generic MacOS menu into winit.
Let me know what you think, if this is something you could see in this repo, then we could collaborate about the details, formats, where to initialize the menu etc.

Screenshot 2020-05-28 at 19 41 00

- [X] Tested on all platforms changed
- [X] Compilation warnings were addressed
- [X] `cargo fmt` has been run on this branch
- [X] `cargo doc` builds successfully
- [ ] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [ ] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [ ] Created or updated an example program if it would help users understand this functionality
- [ ] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

Fixes #41.

### CODE REVIEW

**Comment by casperstorm** [comment]
@kchibisov, you mentioned you had some ideas here. Do you mind sharing them here?

**Comment by adamnemecek** [comment]
Cool feature. Hope this gets merged!

**Comment by OneSadCookie** [comment]
+1 for this; I won't use winit because winit apps don't behave anything like a standard Mac app, and this is a big step forward.

This is very English-centric, should you perhaps call NSLocalizedString (or if winit has a blessed localization solution, possibly use that) for the menu & item titles?

File: src/platform_impl/macos/mod.rs
Comment (aloucks) [inline src/platform_impl/macos/mod.rs:73]: This doesn't seem to need to interact with winit internals at all. Is there some reason this can't be released as an separate crate?

Comment (kchibisov) [reply src/platform_impl/macos/mod.rs:73]: I guess the primary reason for that is that it's something essential for macOS and other windowing toolkits are having such things in them, like glfw. I'm not sure that extracting to other crate will make sense, given that literally every macOS application you'll be creating will require that crate, thus moving that into winit kind of makes sense, since it's less code to maintain and worry about.

Having to always add dependency to each windowing app in cross platform toolkit like winit for just macOS seems like a bad design to me.

**Comment by OneSadCookie** [comment]
This has been open six months now, what do we need to do to get it merged?

**Comment by kchibisov** [comment]
>This has been open six months now, what do we need to do to get it merged?

I haven't seen macOS maintainers in winit for similar amount of time, so someone with macOS should review it. I have macOS machine now, will try to look into it tomorrow, but I'm not a macOS maintainer.

**Comment by casperstorm** [comment]
I found a error myself where the application is not interactable initially launched. If you switch app, and then back then you can interacte with the menu. `Druid` has fixed it in this commit: https://github.com/linebender/druid/commit/05311d1966f673a44ac12794108261b372e978d5#diff-5d6b8d2cc218bec3a8ee3ef9c4c47c61

We should handle this in this PR before we merge.

**Comment by adamnemecek** [comment]
Also, how exactly can my app observe that a menu item has been clicked? Sorry if this is an obvious question.

**Comment by adamnemecek** [comment]
I think that we need to think about the API. The way it works right now is that `NSMenuItem` has a `selector` and `target` fields and when it's clicked, the system invokes `selector` on `target`. This seems kind of terrible for Rust since this would force me to deal implementing objective-c methods on my Rust objects. 

I think that a better API would be adding a new WindowEvent case, called something like `MenuItem` which will allow the user to handle menu item events just like any other event when they are handling winit events.

The way to do this is adding a new struct 
```rust
pub struct MenuItem {
     name: String,
     ...
}
```

and then add some APIs to Window to manage (add, delete) menu items. When you register a new menu item, winit dynamically adds a new method (named the same as the menu item) to the application delegate. When the application delegate receives the message that the item has been clicked, it simply injects a menu item into the event stream.

**Comment by kchibisov** [comment]
>I think that we need to think about the API. The way it works right now is that NSMenuItem has a selector and target properties and when the menu item is clicked, the system invokes selector on target. This seems kind of terrible for Rust since this would force me to deal implementing objective-c methods on my Rust structs.

The thing is that I'm not sure that there should be any special handling for them, since the handling is always the same for everyone? Like if you do `HideOtherWindows`, the only way to handle is either ignore or call winit method to hide other windows, since the amount of menu items is trivial.

I'm not sure about providing custom API right now, since it seems like a very complicated approach and requiring user to match arbitrary strings isn't rust API either.

Right now I think providing a common menu handler for everything(that's what glfw does for example) seems like a reasonable enough solution.

**Comment by adamnemecek** [comment]
@kchibisov I don't follow. When I register these menu items and one gets clicked, how does is that information delivered to my application?

**Comment by kchibisov** [comment]
>@kchibisov I don't follow. When I register these menu items and one gets clicked, how does is that information delivered to my application?

With the current API it doesn't, macOS just calls a selector you've specified, and I'm not sure if you can generate selectors on the fly, especially if we talk about rust side, however I'm not familiar with macOS API.

@casperstorm do you aware if there's a way to provide a custom selector and identify in that selector from which method you've been called, thus all menu items can share a selector?

**Comment by adamnemecek** [comment]
@kchibisov there is not a simple way. You can always do something like forwardInvocation https://stackoverflow.com/questions/9800293/equivalent-of-ruby-method-missing-in-objective-c-ios

however then you have to know if the method you have been passed is a method from menu item.

**Comment by adamnemecek** [comment]
@kchibisov but yes, you can generate selectors on the fly with [NSSelectorFromString](https://developer.apple.com/documentation/foundation/1395294-nsselectorfromstring?language=objc). The trick is that we would be doing this only on the objective c side, on the rust side, we would just inject a `MenuItem` event into the event stream.

**Comment by adamnemecek** [comment]
Another way would be to use the NSMenuItem tag to allow all menu items to share the same selector. That one might also make sense.

**Comment by OneSadCookie** [comment]
I think it's worth distinguishing between "winit needs working menus to act like a standard mac app" and "winit should provide a general menu interface". The latter is a huge problem, with implications for Windows and Linux as well. The former only implies handling a few standard commands in standard ways.

I don't think we should gate this PR on developing a general cross-platform menu interface.

About, Hide/Show, Quit, can be handled in the standard ways without modifying winit's APIs. The other items can be disabled if they can't be omitted.

**Comment by adamnemecek** [comment]
It doesn't need to be general, it can be conditionally compiled for now, but ideally I should be able to use winit to add new menu items right?

**Comment by OneSadCookie** [comment]
Speaking for myself (but also as someone used to consuming GLFW, SDL, etc) I really don't care about custom menu items. I care about "clicking the menubar shows a menu" and "command-Q works as expected and by default".

Not saying winit shouldn't get a general menu system *in the future*, I just think that this is a critical step forward that doesn't require it, and that will be delayed by trying to get that right.

**Comment by kchibisov** [comment]
@adamncasey if you have a solution(code) to a global problem of menu items for macOS, I'm all ears, but I don't care about custom menu items, especially if you really want them you can add some in downstream without many issues.

**Comment by OneSadCookie** [comment]
What this PR delivers already, is all that GLFW and SDL deliver. Parity there is a great first step, and doesn't preclude doing something better in future.

**Comment by kchibisov** [comment]
>What this PR delivers already, is all that GLFW and SDL deliver. Parity there is a great first step, and doesn't preclude doing something better in future.

That's my point as well, since it does precisely what glfw does, and winit is kind of on par with glfw in what it should do (maybe except some minor things like providing OpenGL/Vulkan surface, but that's it).

**Comment by adamnemecek** [comment]
Btw instead of `let _pool = NSAutoreleasePool::new(nil);` you can use the `autorelease` function. I think that a result, you don't need to be call `autorelease`. Check out this example https://github.com/gfx-rs/metal-rs/commit/ae3733ccf6688fc35fe181fae199f192c001b626#diff-d1b0398de8154e02b9ab104eb1b0206139165ec972a8b62a0378ffc367db79a9L16

**Comment by adamnemecek** [comment]
Also, there's this project called [stella2](https://github.com/yvt/Stella2) that does cross platform menu handling https://github.com/yvt/Stella2/blob/c0524f1a18ad4ee72460a0db0d2bdb9fcb48e3a2/stella2/src/view/global.rs#L16

**Comment by LoganDark** [comment]
I think this PR should be merged as-is. Allowing for custom menus can be done later but this is something that is needed now.

**Comment by casperstorm** [comment]
@kchibisov let me know if there's anything I need to do here.

**Comment by LoganDark** [comment]
> @kchibisov let me know if there's anything I need to do here.

I think they're waiting on i18n of the menu text

File: src/platform_impl/macos/menu.rs
Comment (madsmtm) [inline src/platform_impl/macos/menu.rs:41]: We should add the `Services` menu here (can be done by creating a menu item with an empty menu, and registering that as the [`servicesMenu`](https://developer.apple.com/documentation/appkit/nsapplication/1428608-servicesmenu?language=objc))

**Review by madsmtm — CHANGES_REQUESTED** [review]
I welcome this addition, I think it's something that `winit` should definitely support out of the box!

However, I'm working on a library myself for supporting this, and for that it would be really nice this was an optional feature (either compilation or a setting on the event loop), so `winit` and my custom library don't both try to set the main menu!

Haven't looked at `NSAutoreleasePool` in this review, don't know much about it.

**Comment by LoganDark** [comment]
> so `winit` and my custom library don't both try to set the main menu!

Can't you grab the main menu after the fact and modify it?

**Comment by madsmtm** [comment]
> Can't you grab the main menu after the fact and modify it?

Easily, but would like to avoid adding creating a menu twice, but mostly it's because I'm not sure if changing the main menu after it's been assigned would introduce flicker

**Comment by LoganDark** [comment]
> > Can't you grab the main menu after the fact and modify it?
> 
> Easily, but would like to avoid adding creating a menu twice, but mostly it's because I'm not sure if changing the main menu after it's been assigned would introduce flicker

You won't have to create a second menu if a menu is already assigned. Pretty standard "only create if it doesn't already exist" pattern.

**Comment by madsmtm** [comment]
Sure, I could just support that usecase in my library (which it probably should anyhow), and the flicker, if even present, is probably not that big of a deal; it was just a nicety for me, as a user, to have

**Comment by LoganDark** [comment]
> Sure, I could just support that usecase in my library (which it probably should anyhow), and the flicker, if even present, is probably not that big of a deal; it was just a nicety for me, as a user, to have 😉

It's not uncommon for an application to launch without a menu bar and then take _seconds_ to fill in the categories. Seriously, if your library works fast enough for the delay to be called nothing more than a _flicker_, I'll be impressed.

File: src/platform_impl/macos/menu.rs
Comment (ArturKovacs) [reply src/platform_impl/macos/menu.rs:31]: Why would anyone want to use something else then the process name here? That seems to subvert what most users expect in my opinion.

I think it's fine to get from the environment but `NSProcessInfo::processInfo(nil).processName()`is called three times within this function. I think you should call it once and save it into a variable.

**Comment by ArturKovacs** [comment]
As far as I understand `NSLocalizedString` doesn't help with localizing this. First of all `NSLocalizedString` is a C macro and it doesn't seem to be available in rust. According to the docs, the macro returns what `localizedStringForKey:value:table:` would return on the `mainBundle` with the table set to `nil` but as far as I can tell this doesn't magically translate words. It only translates the things that the developer translated themselves and added to the bundle.

My knowledge is very limited on this and for the most part I don't know what I'm talking about so please correct me if I'm wrong. If I'm right however, then I don't think this PR should be concerned with localization.

**Comment by ArturKovacs** [comment]
@casperstorm if you address @madsmtm's comments than I think I would be happy to merge this PR.

**Comment by casperstorm** [comment]
> @casperstorm if you address @madsmtm's comments then I think I would be happy to merge this PR.

Looking into it. Thanks.

File: src/platform_impl/macos/menu.rs
Comment (casperstorm) [reply src/platform_impl/macos/menu.rs:31]: I agree with @ArturKovacs. I have extracted this into a variables for now.

File: src/platform_impl/macos/menu.rs
Comment (casperstorm) [reply src/platform_impl/macos/menu.rs:117]: Agree, actually thought clippy would catch this since I don't think its idiomatic. Good catch.

File: src/platform_impl/macos/menu.rs
Comment (casperstorm) [reply src/platform_impl/macos/menu.rs:41]: Hm, not sure it make sense if we don't have any services to add?

File: src/platform_impl/macos/menu.rs
Comment (casperstorm) [reply src/platform_impl/macos/menu.rs:19]: Would you prefer we pass in a `NSApp()` instead? A little unsure how to correctly solve this.

File: src/platform_impl/macos/mod.rs
Comment (casperstorm) [reply src/platform_impl/macos/mod.rs:73]: When initializing it here, it crashes.

https://github.com/rust-windowing/winit/blob/611d1e871d3a280aef0a816aacd9d2c320b31b15/src/platform_impl/macos/app_state.rs#L274-L282

**Comment by ArturKovacs** [comment]
@casperstorm I added a commit to this PR, please pull and test to see if it works as expected. If you are fine with the changes I made, I think we can merge this.

**Comment by casperstorm** [comment]
> @casperstorm I added a commit to this PR, please pull and test to see if it works as expected. If you are fine with the changes I made, I think we can merge this.

Thanks, @ArturKovacs. It works as expected here.

**Comment by ArturKovacs** [comment]
Shoot! I completely forgot about the activation policy thing. I removed the line that set the activation policy and it still seems to work fine for me. What was the reason you originally put it there @casperstorm? Does my latest commit still work for you?

**Comment by casperstorm** [comment]
> Shoot! I completely forgot about the activation policy thing. I removed the line that set the activation policy and it still seems to work fine for me. What was the reason you originally put it there @casperstorm? Does my latest commit still work for you?

Last commit works for me. Honestly I can't remember why I put it there. It's been quite sometime.

**Comment by ArturKovacs** [comment]
Thanks for the contribution!

**Comment by OneSadCookie** [comment]
Thank you @casperstorm for doing the work and @ArturKovacs for getting it merged! Been waiting a long time for this one :)

**Pull request merged**
